### ISSUE TITLE
Backdriving against large torque causes jerky torque reduction

### ISSUE THREAD

**Issue by dzid26** [dzid26/StepperServoCAN#15]
With high torque (high current) and backdriving against it, there are two effects contributing to the jerky movement. 
1. In the case of the present current control, PWM to DC filters cause DC offset  and more crucially phase delay on the current request. Partially reduced by [this](https://github.com/dzid26/StepperServoCAN/blob/dc3cfa9c9782628a55a49ea95079d300faa36db6/firmware/src/BSP/A4950.c#L38).
2. High motor inductance causes phase delay on the current request

When backdriving under smaller currents, both effects are not as visible as the current can quickly change direction. For larger currents, it takes time for it to change - the phase delay.

The phase delay can be compensated using speed calculation.

**Comment by OxygenLiu** [comment]
The present Sine commutation may lose steps when backdriving. SimpleFOC implements a FOC library for closed loop control of stepper motors, as demonstrated here https://youtu.be/zcb86TRxTxc

**Comment by dzid26** [comment]
SimpleFOC has all kind of control schemes. I like [voltage control](https://docs.simplefoc.com/voltage_torque_mode#voltage-control-with-current-estimation-and-back-emf-compensation) the most for our board. 
For full-blown FOC you would need different hardware with inline current sensing and fast phase current close loop control done by the MCU.

Current [hardware](https://github.com/dzid26/StepperServo-hardware/files/11010020/StepperServoCAN-schematic.pdf) is simpler, but I believe this issue can still be solved by using knowledge of motor inductance and sensor delays.

**Comment by dzid26** [comment]
Currents data during the snap event:
[Snap on overload.xlsx](https://github.com/dzid26/StepperServoCAN/files/13233195/Snap.on.overload.xlsx)

Current changes fast during the snap, because the shaft accelerates rapidly (jerks, snaps).
<img width="904" alt="image" src="https://github.com/dzid26/StepperServoCAN/assets/841061/2e08591f-39ae-4c2d-b0b7-ba12a8950ab5">
But why does it snap? Either current is not commanded right or it has something to do with large torque ripple under load. (#11)  To be investigated...

**Comment by OxygenLiu** [comment]
looks like some variable is reset or overflow?

**Comment by dzid26** [comment]
I don't think so. 
The current waveworms draw a circle, which is impressive given the rate of change of things.  

But It could be that the timing is not ideal - but things don't happen fast before the snap occurs.

I am leaning toward that this could be hardware characteristics - the large torque ripple under high load coupled with springiness in the load in my setup and non-linear friction (stiction) in the case of a gearbox.  

I think fully solving #11 would fix this too.  I would only hope that the torque peaks are aligned at multiples of 90deg and not 45, because in one case we can extract more useful torque and power from the motor, and in the other, we may lose average power.

**Comment by OxygenLiu** [comment]
Right, I observe somewhat larger output/stall torque if the stepper can spin up a bit. It performs a bit worse with rigid or springless connection to the load.  This is only my general impression without detailed measurements. 

Probably the current snap is the transition phenomena  in `slew range` between  `pull in`  and `pull out` regions. [Ref Link](https://www.linearmotiontips.com/stepper-motor-torque-speed-characteristics-explained/)

![](https://www.linearmotiontips.com/wp-content/uploads/2019/09/Torque-Curves.png)

**Comment by OxygenLiu** [comment]
The distortion is probably caused by the slow/mixed decay mode,  TI proposes the dynamic decay mode in smart tune. 

[TI Smart Tune](https://youtu.be/8i_HDJmxais?si=hYQXFomMoEOjRZmi&t=134)

**Comment by dzid26** [comment]
I think I mixed two issues/scenarios here. :(

1. One could be related to decay modes and distortion - which at higher speed oposing the driving torque, the torque value drops. But it is normal scenario of driving the moror when bemf makes it harder to maintain the high current. Here we are **braking**, so bemf helps to have high braking current. 

2. At much higher requested torque (say 3Amp) somewhat similar effect can be visible at 0 speed! If the load torque increases, eventually the motor is "giving up" and the torque drops. It's the snap I mentioned last week. 

However I think underlying causes are different. Although eventually they may get combined once the speed increases. 

For the 2. scenario, (which is more disturbing), the decay mode doesn't play a role. -  initially the speed is close to zero (before it snaps and looses torque), and so bemf is close to zero and slow decay will do just fine. It is confirmed by the circle plot above. 

Realistically, I think the seconds scenario is caused by nonlinearity if the torque - which in turn is probably amplfied by localized magnetic saturation during high current. - That is the holding torque is probably only high where teeth are aligned and relatively small in-between. (Or opposite way). And so once I go pass the tooth, I loose torque and motor speed goes up due to inertia and springiness of the system. Once it starts rotating, perhaps the torque never comes back due to combination of the 1. scenario.
